services/web/server/src/simcore_service_webserver/login/handlers_confirmation.py

```diff
@@ -54,6 +66,52 @@ class _PathParam(BaseModel):
     code: SecretStr
 
 
+def _parse_extra_credits_or_none(
+    confirmation: ConfirmationTokenDict,
+) -> PositiveInt | None:
+    with suppress(ValidationError, JSONDecodeError):
+        invitation = InvitationData.parse_raw(confirmation.get("data", "EMPTY"))
+        return invitation.extra_credits
+    return None
+
+
+async def _handle_confirm_registration(
+    app: web.Application,
+    product_name: ProductName,
+    confirmation: ConfirmationTokenDict,
+):
+    db: AsyncpgStorage = get_plugin_storage(app)
+    user_id = confirmation["user_id"]
+
+    # activate user and consume confirmation token
+    await db.delete_confirmation_and_update_user(
```

Comment:
Q: just a thought, would it not be better to delete after the user is notified?

In case of error with the notification, if the user retries the invitation will have already been used. Is this the behaviour we want?
If this is the case, I would delete it immediately after the call to `validate_confirmation_code`.

Reply:
the notification function below is not to notify users. It is to trigger an event "ON_USER_CONFIRMED". The wallet plugin responds to that event by adding a primary wallet and credits to it (if invitation contains it)